lua/neogit/watcher.lua

```diff
@@ -0,0 +1,85 @@
+local M = {}
+
+local uv = vim.loop
+
+local config = require("neogit.config")
+local logger = require("neogit.logger")
+local util = require("neogit.lib.util")
+local status = require("neogit.status")
+local git = require("neogit.lib.git")
+
+local Path = require("plenary.path")
+local a = require("plenary.async")
+
+M.watcher = {}
+
+local function git_dir()
+  return Path.new(require("neogit.lib.git").repo.git_root, ".git"):absolute()
+end
+
+function M.setup()
+  local gitdir = git_dir()
+  local watcher = M.watch_git_dir(gitdir)
+
+  M.watcher[gitdir] = watcher
+end
+
+-- Adapted from https://github.com/lewis6991/gitsigns.nvim/blob/main/lua/gitsigns/manager.lua#L575
+--- @param gitdir string
+--- @return uv_fs_event_t?
+function M.watch_git_dir(gitdir)
+  if not config.values.auto_refresh then
+    return
+  end
+
+  if M.watcher[gitdir] then
+    logger.debug(string.format("[WATCHER] for '%s' already setup! Bailing.", gitdir))
+    return
+  end
+
+  local watch_gitdir_handler_db = util.debounce_trailing(
+    200,
+    a.void(function()
+      logger.debug("[WATCHER] Dispatching Refresh")
+      git.repo:dispatch_refresh { callback = status.dispatch_refresh, source = "watcher" }
+    end)
+  )
+
+  logger.debug("[WATCHER] Watching git dir: " .. gitdir)
+
+  local w = assert(uv.new_fs_event())
+
+  w:start(
+    gitdir,
+    {},
```

Comment:
Should this be recursive?

I don't think this will catch changes in `.git/index/file` etc, though I may be wrong

http://docs.libuv.org/en/v1.x/guide/filesystem.html#file-change-events

https://github.com/luvit/luv/blob/master/docs.md#uvfs_event_startfs_event-path-flags-callback

Reply:
Yeah, it's recursive, no worries about that

Reply:
Huh, maybe I'm mistaken. Not passing the recursive flag.

```
                - `flags`: `table`
                  - `watch_entry`: `boolean` or `nil` (default: `false`)
                  - `stat`: `boolean` or `nil` (default: `false`)
                  - `recursive`: `boolean` or `nil` (default: `false`)
```